### PR TITLE
Graphql helper

### LINKED ISSUE #1723: Helper functions for GraphQL graph mutations
Need to add helper functions to the client for the following operations:

- [ ] new_graph
- [ ] client.graph(path=).add_node()
- [ ] client.graph(path=).add_nodes()
- [ ] client.graph(path=).add_edge()
- [ ] client.graph(path=).add_edges()
- [ ] client.graph(path=).delete_edge()
- [ ] client.graph(path=).add_properties()
- [ ] client.graph(path=).add_constant_properties()
- [ ] client.graph(path=).update_constant_properties()
- [ ] client.graph(path=).node(name=).add_constant_properties()
- [ ] client.graph(path=).node(name=).set_node_type()
- [ ] client.graph(path=).node(name=).update_constant_properties()
- [ ] client.graph(path=).node(name=).add_updates()
- [ ] client.graph(path=).edge(src=,dst=).delete()
- [ ] client.graph(path=).edge(src=,dst=).add_constant_properties()
- [ ] client.graph(path=).edge(src=,dst=).update_constant_properties()
- [ ] client.graph(path=).edge(src=,dst=).add_updates()

### PR DESCRIPTION
**Pull request by miratepuffin**
### What changes were proposed in this pull request?

- Tided up graphql pytests as having a single file was driving me mad
- Broke the graphql python module into smaller files
- Exposed `delete` on edge in python
- Fixed missing DTime parsing on Python properties
- Added RemoteGraph, RemoteNode and RemoteEdge classes to wrap updates to graph on a server
- Added below functions with tests to these classes
- Made Properties options in add_updates for Node and Edge in Graphql
- Added some helper functions in make for formatting rust/python, building docs and building python
- fixed a bug in batch add nodes, where you couldn't just add the node_type.
- Fixed a bug in import_node(s)/import_edge(s) when the graph is indexed with ints
## Functions added
- [x] new_graph
- [x] client.graph(path=).add_node()
- [x] client.graph(path=).add_nodes()
- [x] client.graph(path=).add_edge()
- [x] client.graph(path=).add_edges()
- [x] client.graph(path=).delete_edge()
- [x] client.graph(path=).add_properties()
- [x] client.graph(path=).add_constant_properties()
- [x] client.graph(path=).update_constant_properties()
- [x] client.graph(path=).node(name=).add_constant_properties()
- [x] client.graph(path=).node(name=).set_node_type()
- [x] client.graph(path=).node(name=).update_constant_properties()
- [x] client.graph(path=).node(name=).add_updates()
- [x] client.graph(path=).edge(src=,dst=).delete()
- [x] client.graph(path=).edge(src=,dst=).add_constant_properties()
- [x] client.graph(path=).edge(src=,dst=).update_constant_properties()
- [x] client.graph(path=).edge(src=,dst=).add_updates()

### Why are the changes needed?
Dealing with graphql strings in python is a pain

### Does this PR introduce any user-facing change? If yes is this documented?
yes and yes

### How was this patch tested?
New tests added for each function 

### Issues
fixes #1723

### CODE REVIEW

File: raphtory/src/python/types/wrappers/prop.rs [hunk 1]
```diff
@@ -73,6 +73,9 @@ impl<'source> FromPyObject<'source> for Prop {
         if let Ok(d) = ob.extract() {
             return Ok(Prop::NDTime(d));
         }
+        if let Ok(d) = ob.extract() {
+            return Ok(Prop::DTime(d));
```
Comment (miratepuffin) [inline raphtory/src/python/types/wrappers/prop.rs:77]: Can we use our other Datetime parser here

Comment (miratepuffin) [reply raphtory/src/python/types/wrappers/prop.rs:77]: Will make this a seperate ticket

File: raphtory-graphql/src/python/client/remote_node.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    /// Set the type on the node. This only works if the type has not been previously set, otherwise will
+    /// throw an error
+    ///
+    /// Parameters:
+    ///     new_type (str): The new type to be set
+    ///
+    /// Returns:
+    ///     Result: A result object indicating success or failure.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_node.rs:37]: No result in python

File: raphtory-graphql/src/python/client/remote_node.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    /// This function allows for the addition of property updates to a node within the graph. The updates are time-stamped, meaning they are applied at the specified time.
+    ///
+    /// Parameters:
+    ///     t (PyTime): The timestamp at which the updates should be applied.
+    ///     properties (Optional[Dict[str, Prop]]): A dictionary of properties to update.
+    ///         Each key is a string representing the property name, and each value is of type Prop representing the property value.
+    ///         If None, no properties are updated.
+    ///
+    /// Returns:
+    ///     Result: A result object indicating success or failure.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_node.rs:70]: No result in python

File: raphtory-graphql/src/python/client/remote_node.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        let query = build_query(template, query_context)?;
+        let _ = &self.client.query(py, query, None)?;
+        Ok(())
+    }
+
+    /// Add updates to a node in the remote graph at a specified time.
+    /// This function allows for the addition of property updates to a node within the graph. The updates are time-stamped, meaning they are applied at the specified time.
+    ///
+    /// Parameters:
+    ///     t (PyTime): The timestamp at which the updates should be applied.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_node.rs:64]: `PyTime` should be `(int | str | DateTime)`

File: raphtory-graphql/src/python/client/remote_node.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    /// This function is used to add properties to a node that remain constant and does not
+    /// change over time. These properties are fundamental attributes of the node.
+    ///
+    /// Parameters:
+    ///     properties (Dict[str, Prop]): A dictionary of properties to be added to the node.
+    ///     Each key is a string representing the property name, and each value is of type Prop
+    ///     representing the property value.
+    ///
+    /// Returns:
+    ///     Result: A result object indicating success or failure.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_node.rs:110]: No result in python

File: raphtory-graphql/src/python/client/remote_node.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    /// This function is used to add properties to a node that remain constant and do not
+    /// change over time. These properties are fundamental attributes of the node.
+    ///
+    /// Parameters:
+    ///     properties (Dict[str, Prop]): A dictionary of properties to be added to the node.
+    ///     Each key is a string representing the property name, and each value is of type Prop
+    ///     representing the property value.
+    ///
+    /// Returns:
+    ///     Result: A result object indicating success or failure.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_node.rs:147]: No result in python

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+impl PyRemoteGraph {
+    #[new]
+    pub(crate) fn new(path: String, client: PyRaphtoryClient) -> Self {
+        Self { path, client }
+    }
+
+    /// Gets a remote node with the specified id
+    ///
+    /// Arguments:
+    ///   id (str or int): the node id
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:34]: `(str | int)`

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+    ///   RemoteNode
+    pub fn node(&self, id: GID) -> PyRemoteNode {
+        PyRemoteNode::new(self.path.clone(), self.client.clone(), id.to_string())
+    }
+
+    /// Gets a remote edge with the specified source and destination nodes
+    ///
+    /// Arguments:
+    ///     src (str or int): the source node id
+    ///     dst (str or int): the destination node id
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:46]: `str | int`

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+            dst.to_string(),
+        )
+    }
+
+    /// Batch add node updates to the remote graph
+    ///
+    /// Arguments:
+    ///   updates (List[RemoteNodeAddition]): The list of updates you want to apply to the remote graph
+    /// Returns:
+    ///   None
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:65]: not needed

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+
+        Ok(())
+    }
+
+    /// Batch add edge updates to the remote graph
+    ///
+    /// Arguments:
+    ///   updates (List[RemoteEdgeAddition]): The list of updates you want to apply to the remote graph
+    /// Returns:
+    ///   None
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:134]: not needed

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+        let query = build_query(template, query_context)?;
+        let _ = &self.client.query(py, query, None)?;
+
+        Ok(())
+    }
+
+    /// Adds a new node with the given id and properties to the remote graph.
+    ///
+    /// Arguments:
+    ///    timestamp (int, str, or datetime(utc)): The timestamp of the node.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:202]: `(int | str | DateTime)`

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+    ///    id (str or int): The id of the node.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:203]: `(int | str)`

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+    ///    properties (dict): The properties of the node (optional).
+    ///    node_type (str): The optional string which will be used as a node type
+    /// Returns:
+    ///   the added node (RemoteNode)
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:207]: `RemoteNode: the added node`

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+            self.path.clone(),
+            self.client.clone(),
+            id.to_string(),
+        ))
+    }
+
+    /// Adds properties to the remote graph.
+    ///
+    /// Arguments:
+    ///    timestamp (int, str, or datetime(utc)): The timestamp of the temporal property.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:249]: see above

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+    ///
+    /// Arguments:
+    ///    timestamp (int, str, or datetime(utc)): The timestamp of the edge.
+    ///    src (str or int): The id of the source node.
+    ///    dst (str or int): The id of the destination node.
+    ///    properties (dict): The properties of the edge, as a dict of string and properties (optional).
+    ///    layer (str): The layer of the edge (optional).
+    ///
+    /// Returns:
+    ///   The added edge (RemoteEdge)
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:352]: see above

File: raphtory-graphql/src/python/client/remote_graph.rs [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+    /// Deletes an edge in the remote graph, given the timestamp, src and dst nodes and layer (optional)
+    ///
+    /// Arguments:
+    ///   timestamp (int): The timestamp of the edge.
+    ///   src (str or int): The id of the source node.
+    ///   dst (str or int): The id of the destination node.
+    ///   layer (str): The layer of the edge. (optional)
+    ///
+    /// Returns:
+    ///  The deleted edge (RemoteEdge)
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/client/remote_graph.rs:401]: see above

File: raphtory-graphql/src/python/server/server.rs [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+    /// They have to be specified using the `input` parameter as a dict where the keys are the
+    /// names of the parameters and the values are the types, expressed as strings.
+    ///
+    /// Arguments:
+    ///   * `name` (`str`): the name of the function in the GraphQL schema.
+    ///   * `input` (`dict`): the keyword arguments expected by the function.
+    ///   * `function` (`function`): the function to run.
+    ///
+    /// Returns:
+    ///    A new server object containing the vectorised graphs.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/server/server.rs:233]: this is not correct python format, no `` ` `` or `*` needed

File: raphtory-graphql/src/python/server/server.rs [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+    ///   If no embedding function is provided, the server will attempt to use the OpenAI API
+    ///   embedding model, which will only work if the env variable OPENAI_API_KEY is set
+    ///   appropriately
+    ///
+    /// Arguments:
+    ///   * `graph_names`: the names of the graphs to vectorise. All by default.
+    ///   * `cache`: the directory to use as cache for the embeddings.
+    ///   * `embedding`: the embedding function to translate documents to embeddings.
+    ///   * `node_document`: the property name to use as the source for the documents on nodes.
+    ///   * `edge_document`: the property name to use as the source for the documents on edges.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/server/server.rs:181]: this is not correct python format, no `` ` `` or `*` needed

File: raphtory-graphql/src/python/server/server.rs [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+    ///
+    /// The function needs to take a `GraphqlGraphs` object as the first argument followed by a
+    /// pre-defined set of keyword arguments. Supported types are `str`, `int`, and `float`.
+    /// They have to be specified using the `input` parameter as a dict where the keys are the
+    /// names of the parameters and the values are the types, expressed as strings.
+    ///
+    /// Arguments:
+    ///   * `name` (`str`): the name of the function in the GraphQL schema.
+    ///   * `input` (`dict`): the keyword arguments expected by the function.
+    ///   * `function` (`function`): the function to run.
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/server/server.rs:255]: this is not correct python format, no `` ` `` or `*` needed

File: raphtory-graphql/src/python/server/server.rs [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+            PyGlobalPlugins(entry_point.clone()).into_py(py)
+        };
+        PyGraphServer::with_generic_document_search_function(slf, name, input, function, adapter)
+    }
+
+    /// Start the server and return a handle to it.
+    ///
+    /// Arguments:
+    ///   * `port`: the port to use (defaults to 1736).
+    ///   * `timeout_ms`: wait for server to be online (defaults to 5000). The server is stopped if not online within timeout_ms but manages to come online as soon as timeout_ms finishes!
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/server/server.rs:275]: this is not correct python format, no `` ` `` or `*` needed

File: raphtory-graphql/src/python/server/server.rs [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+            }
+        } else {
+            Err(PyException::new_err("Failed to start server"))
+        }
+    }
+
+    /// Run the server until completion.
+    ///
+    /// Arguments:
+    ///   * `port`: the port to use (defaults to 1736).
```
Comment (ljeub-pometry) [inline raphtory-graphql/src/python/server/server.rs:333]: this is not correct python format, no `` ` `` or `*` needed

File: raphtory/src/python/graph/edge.rs [hunk 22]
```diff
@@ -363,6 +364,18 @@ impl PyMutableEdge {
             .add_updates(t, properties.unwrap_or_default(), layer)
     }
 
+    /// Mark the edge as deleted at the specified time.
+    ///
+    /// Parameters:
+    ///     t (PyTime): The timestamp at which the deletion should be applied.
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:370]: `(int | str | DateTime)`

File: raphtory/src/python/graph/edge.rs [hunk 23]
```diff
@@ -363,6 +364,18 @@ impl PyMutableEdge {
...
+    ///     layer (str): The layer you want the deletion applied to .
+    ///
+    /// Returns:
+    ///     Result: A result object indicating success or failure.
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:374]: no result in python

File: raphtory/src/python/graph/graph.rs [hunk 24]
```diff
@@ -169,7 +169,7 @@ impl PyGraph {
     ///    properties (dict): The properties of the node (optional).
     ///    node_type (str): The optional string which will be used as a node type
     /// Returns:
-    ///   None
+    ///   the added node (Node)
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/graph.rs:172]: `Node: ...`

File: raphtory/src/python/graph/graph.rs [hunk 25]
```diff
@@ -236,7 +236,7 @@ impl PyGraph {
     ///    layer (str): The layer of the edge (optional).
     ///
     /// Returns:
-    ///   None
+    ///   returns the added Edge (Edge)
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/graph.rs:239]: `Edge: ...`

**Pull request merged**